wptrunner: Fail with NotImplementedError when appropriate

nsatragno opened the pull request:
Modify the base executor so that if a certain protocol message is not
implemented, we fail with NotImplementedError instead of raising an
exception and crashing the browser. Have the virtual authenticator API
explicitly throw this error on marionette.

nsatragno committed:
Fail nicely with NotImplementedError when API not implemented

Modify the base executor so that if a certain protocol message is not
implemented, we fail with NotImplementedError instead of raising an
exception and crashing the browser. Have the virtual authenticator API
explicitly throw this error on marionette.

nsatragno commented:
@Hexcles please take a look, thanks! ^_^

Hexcles commented on tools/wptrunner/wptrunner/executors/base.py:
```diff
@@ -631,6 +631,9 @@ def process_action(self, url, payload):
             result = action_handler(payload)
+        except NotImplementedError:
+            self.logger.warning("Action %s not implemented" % action)
+            self._send_message("complete", "not-implemented")
```
I'm not so sure about this. It looks like `not-implemented` will be a new status. Do we need to define it somewhere?

A less risky way is to use the existing `error` status, and put "action %s not implemented" into the message field.

nsatragno replied:
> A less risky way is to use the existing error status, and put "action %s not implemented" into the message field.

Done.

Hexcles reviewed (commented):
If I understand the existing code correctly, this is a good idea.

We are talking about a WebDriver extension here. If the WebDriver executor is used but the underlying WebDriver implementation does not support this extension, we'd see a protocol error nicely handled, but in the case of the Marionette executor, the Python code would explode, generating an internal error I believe, and the browser gets killed.

Hexcles commented:
@jgraham as this is Marionette-specific

nsatragno committed:
Send generic error instead and add message.

Also update expectations.

nsatragno committed:
Reject promise with the correct error on fail.

gsnedders reviewed (commented):
On the face of it, what's the issue with tests just ERRORing in other browsers? We'll still ERROR in general, just reject the promise (and FAIL) in the specific case of Marionette and `VirtualAuthenticatorProtocolPart`. If we want to return an error message back to the browser for unsupported actions, I'd rather we do it for all.

Hexcles commented:
Yeah I agree. The end result (according to the expectation) is a bit weird on Firefox: the test-level status is ERROR but all subtests fail.

It'd be great to avoid those FAIL and make all browsers consistent. I think the goal here is to avoid the internal error that'd kill the browse, but we don't want to make e.g. Firefox special, either. Do other browsers get killed when calling unknown commands? For example on Safari we are seeing an `UnknownCommandException`: https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=37203&view=logs&jobId=c092c851-ddde-53a1-8da2-b0c5715821df&taskId=86ea80b8-d5d9-55d3-a57c-ef30822b0c90&lineStart=496&lineEnd=497&colStart=1&colEnd=1 but I'm not sure if that kills the browser or not.

nsatragno committed:
Throw unimplemented for webdriver as well

When a webdriver method returns 404, bubble the error all the way to the
javascript test as well.

Also add NotImplementedError for Marionette's generate_test_report
method.

nsatragno commented:
> On the face of it, what's the issue with tests just ERRORing in other browsers?

Bubbling the error all the way back to the js test as a promise rejection has two advantages: it allows browsers to implement features partially and see some subtests pass and others fail and it allows for tests to make use of automation when available: try using an automation API and if it isn't implemented, default to a manual action. This latter case is the reason I'm specifically interested in making this change, the webauthn tests are in sore need of automation but I don't want to break them for browsers that don't support the automation API yet :)

>  in the specific case of Marionette and VirtualAuthenticatorProtocolPart. If we want to return an error message back to the browser for unsupported actions, I'd rather we do it for all.

Okay, added it for the reporting API. I think that's the only other protocol part not supported by marionette. I don't know how to implement this in a more generic way without hijacking the AttributeError exception that is currently thrown for this case.

> It'd be great to avoid those FAIL and make all browsers consistent. I think the goal here is to avoid the internal error that'd kill the browse, but we don't want to make e.g. Firefox special, either. Do other browsers get killed when calling unknown commands? For example on Safari we are seeing an UnknownCommandException

The `UnknownCommandException` is treated as any other `Exception` by `base.py`. Catching it and treating it same as `NotImplementedError` does the trick and standardizes the behaviour among executors. Updated the PR to do that. Unfortunately it also makes the unittest import fail for some reason, and I have no idea how to fix that :(

nsatragno committed:
Add localpaths to wptrunner test.

nsatragno committed:
Silence unused import error

Hexcles commented:
Safari failed the infrastructure test unexpectedly:

```
/infrastructure/testdriver/virtual_authenticator.html
  ERROR /infrastructure/testdriver/virtual_authenticator.html - authenticator_id
Traceback (most recent call last):
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 292, in run_func
    self.result = True, self.func(self.protocol, self.url, self.timeout)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/executorwebdriver.py", line 385, in do_testharness
    done, rv = handler(result)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 676, in __call__
    return callback(url, payload)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 690, in process_action
    result = action_handler(payload)
  File "/Users/runner/runners/2.160.1/work/1/s/tools/wptrunner/wptrunner/executors/base.py", line 794, in __call__
    authenticator_id = payload["authenticator_id"]
KeyError: 'authenticator_id'
```

I am not entirely sure what's going on. I suspect this has something to do with the usage of object destructuring in https://github.com/web-platform-tests/wpt/blob/ca4e480f83ee0dbae094ec88a7a8e5851b17af93/tools/wptrunner/wptrunner/testdriver-extra.js#L138

*Perhaps* if the action is not supported, we get a null/undefined from `add_virtual_authenticator`, which then end up causing the subsequent payloads to miss the `authenticator_id` field.

We could use an explicit declaration in the front-end, or change the Python code to use `.get`: https://github.com/web-platform-tests/wpt/blob/ca4e480f83ee0dbae094ec88a7a8e5851b17af93/tools/wptrunner/wptrunner/executors/base.py#L727

nsatragno committed:
Initialize authenticator_id.

nsatragno commented:
> Safari failed the infrastructure test unexpectedly:

The reason the test is failing is we were not initializing `authenticator_id` on the [infrastructure test](https://github.com/web-platform-tests/wpt/pull/20409/files#diff-bc78bb51ee567722165ea30eb8090462R27), so when `add_virtual_authenticator` failed it never got set which triggered the `KeyError` on the next test.

> We could use an explicit declaration in the front-end, or change the Python code to use .get:

I decided to simply initialize `authenticator_id` to `null` which _should hopefully_ avoid this problem and keep the virtual authenticator wptrunner code consistent with the other endpoints. What do you think?

Hexcles commented on tools/wptrunner/wptrunner/testdriver-extra.js:
```diff
@@ -19,7 +19,7 @@
             pending_resolve(result);
         } else {
-            pending_reject();
+            pending_reject({ status: data.status, message: data.message });
```
Reading the logs, I realize this isn't very human-readable, e.g.

`promise_test: Unhandled rejection with value: object "[object Object]"`

We probably want to return a simple string instead, something like: `${data.status}: ${data.message}`

nsatragno replied:
Agreed, done!

Hexcles reviewed (commented):
```
 1:45.30 TEST_START: /infrastructure/testdriver/set_permission.html
 1:45.53 WARNING Action set_permission not implemented
 1:45.54 WARNING Action set_permission not implemented
 1:45.57 TEST_END: Test OK, expected ERROR. Subtests passed 0/2. Unexpected 2
FAIL Grant Permission for one realm - promise_test: Unhandled rejection with value: object "[object Object]"
promiseReactionJob@[native code]
FAIL Deny Permission, omit one realm - promise_test: Unhandled rejection with value: object "[object Object]"
promiseReactionJob@[native code]
UNEXPECTED-OK /infrastructure/testdriver/set_permission.html
```

So we also need a new expectation for `/infrastructure/testdriver/set_permission.html` on Safari.

nsatragno committed:
Add SetPermissionProtocol and rebaseline expectations.

nsatragno commented:
This patch is now blocked on #20603

nsatragno committed:
Add more descriptive error message.

Hexcles commented:
> This patch is now blocked on #20603

It has been merged. I just restarted the

Hexcles reviewed (approved):
Hooray! Finally, everything is green!

This change now looks good to me. Pinging @jgraham again as this changes the Marionette executor and it'd probably change some expectations in Gecko.

nsatragno commented:
Rebased -- ptal @jgraham

jgraham reviewed (approved):
This seems reasonable to me.
